### PR TITLE
handle unpaired read data

### PR DESCRIPTION
**Pull request by dpark01**
This should mostly address issue #238. Unit tests for single-end reads have not been written, though some small data sets have been tried by hand, and existing tests on paired and empty inputs are unbroken.

This also includes some cleanup of legacy fastq commands in various places, and some moving around of code.

The following steps were not previously working on single-end read data: bmtagger, blastn, lastal, trimmomatic, novoalign

### CODE REVIEW

File: tools/mvicuna.py [hunk 1]
```diff
@@ -63,6 +63,34 @@ def rmdup(self, inPair, outPair, outUnpaired=None):
...
 
+    def rmdup_single(self, inFastq, outFastq):
+        """
+        Run mvicuna's duplicate removal operation on single-end input reads in
+            fastq format, producing output in fastq format.
+        Notes on weird behaviors of M-Vicuna DupRm:
+            For some reason, it requires you to specify both -opfq and -drm_op even for
+            single-end input.
+
+            Actually, more interestingly, this doesn't work at all.
```
Comment (tomkinsc) [inline tools/mvicuna.py]: DumpRm does not work at all on single-end reads?

File: tools/novoalign.py [hunk 2]
```diff
@@ -177,7 +177,7 @@ def align_one_rg_bam(self, inBam, refFasta, outBam, rgid=None, options=None, min
         tmp_sam = util.file.mkstempfname('.novoalign.sam')
         tmp_sam_err = util.file.mkstempfname('.novoalign.sam.err')
         cmd = [self.install_and_get_path(), '-f', one_rg_inBam] + list(map(str, options))
-        cmd = cmd + ['-F', 'BAMPE', '-d', self._fasta_to_idx_name(refFasta), '-o', 'SAM']
+        cmd = cmd + ['-F', 'BAM', '-d', self._fasta_to_idx_name(refFasta), '-o', 'SAM']
```
Comment (tomkinsc) [inline tools/novoalign.py]: We should probably have a comment here noting that the use of `-F BAM` is possible via the newer Novoalign version, since it does not seem to be mentioned in the current reference manual.

File: tools/samtools.py [hunk 3]
```diff
@@ -69,7 +69,7 @@ def view(self, args, inFile, outFile, regions=None):
 
     def bam2fq(self, inBam, outFq1, outFq2=None):
         if outFq2 is None:
-            self.execute('bam2fq', ['-s', outFq1, inBam])
+            self.execute('bam2fq', ['-n', inBam], stdout=outFq1)
```
Comment (tomkinsc) [inline tools/samtools.py]: Why the change to use stdout over `-s <file>`? Does it conflict with `-n`?

File: tools/trimmomatic.py [hunk 4]
```diff
@@ -15,6 +20,58 @@ def __init__(self, install_methods=None):
...
+        '''Trim read sequences with Trimmomatic.'''
+        trimmomaticPath = self.install_and_get_path()
+        unpairedFastq1 = unpairedOutFastq1 or util.file.mkstempfname()
+        unpairedFastq2 = unpairedOutFastq2 or util.file.mkstempfname()
+        javaCmd = [trimmomaticPath]
+
+        if inFastq2 is None or os.path.getsize(inFastq2) < 10:
+            # Unpaired reads
+            javaCmd.extend([
+                    'SE', '-phred33',
```
Comment (tomkinsc) [inline tools/trimmomatic.py]: Should we expose this (and the `PE` quality encoding setting) as a parameter?

File: tools/trimmomatic.py [hunk 5]
```diff
@@ -15,6 +20,58 @@ def __init__(self, install_methods=None):
...
+        else:
+            # Paired reads
+            javaCmd.extend([
+                    'PE', '-phred33',
+                    inFastq1, inFastq2,
+                    pairedOutFastq1, unpairedFastq1, pairedOutFastq2, unpairedFastq2
+                ])
+
+        # all the options
+        javaCmd.extend(
```
Comment (tomkinsc) [inline tools/trimmomatic.py]: Should we explicitly expose the ability to set `-threads <int>`? As per the [manual](http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf), "If not specified, it will be chosen automatically." On the cluster this could be an issue if it detects the number of system cores rather than the number of cores available to the job.

File: read_utils.py [hunk 6]
```diff
@@ -408,53 +408,8 @@ def main_filter_bam(args):
 
 __commands__.append(('filter_bam', parser_filter_bam))
 
-# =======================
-# ***  bam_to_fastq   ***
-# =======================
-
-
-def bam_to_fastq(
```
Comment (tomkinsc) [inline read_utils.py]: The argument for removal is that users can simply use samtools' `bam2fq`?

File: assembly.py [hunk 7]
```diff
@@ -58,7 +58,7 @@ def __init__(self, n_start, n_trimmed, n_rmdup, n_output, n_subsamp, n_unpaired_
 
 def trim_rmdup_subsamp_reads(inBam, clipDb, outBam, n_reads=100000):
```
Comment (tomkinsc) [inline assembly.py]: No issues in this function with single-end reads?

File: tools/novoalign.py (hunk 2 above)
Comment (dpark01) [inline tools/novoalign.py]: weirdly, if you google for "Novoalign manual", there are a few top hits referring to different versions. The one for [v3.07](http://www.novocraft.com/userfiles/file/Novocraft.pdf) describes this. The one for [v3.02](http://www.novocraft.com/wp-content/uploads/Novocraft.pdf) does not.

File: tools/mvicuna.py (hunk 1 above)
Comment (dpark01) [inline tools/mvicuna.py]: I tried valiantly... maybe a bug or something, maybe it was never tested in that scenario. I figure a nop is better than a failure.

File: tools/trimmomatic.py (hunk 4 above)
Comment (dpark01) [inline tools/trimmomatic.py]: I think I prefer autodetection where possible, partly because this is really a fastq-specific problem, and the user's input files (BAM) can always be paired, single, or a mixture of both. If it's a mixture, I wouldn't know how to expose the parameter to the user.

File: read_utils.py (hunk 6 above)
Comment (dpark01) [inline read_utils.py]: Yeah, and also I'm just starting to clean out a few exposed things that are not used in the pipeline (from the user's point of view).

File: assembly.py (hunk 7 above)
Comment (dpark01) [inline assembly.py]: Once you fix trimmomatic, apparently not! It's kind of beautiful, especially watching the little report at the end with all the counts.

File: tools/novoalign.py (hunk 2 above)
Comment (tomkinsc) [inline tools/novoalign.py]: I see...serves me right for assuming the version linked from the Novocraft docs website would be the most recent version.

File: tools/trimmomatic.py (hunk 4 above)
Comment (tomkinsc) [inline tools/trimmomatic.py]: Ok, makes sense!
